lib/texture2.c

```diff
@@ -246,8 +246,8 @@ ktxFormatSize_initFromDfd(ktxFormatSize* This, ktx_uint32_t* pDfd)
     }
     // pDfd[0] contains the size in words as per createdfd.c writeHeader()
     // Check if DFD buffer is large enough to hold all the declared samples
-    uint32_t samplesSize = sizeof(uint32_t) * (1 + KHR_DFDSIZEWORDS(KHR_DFDSAMPLECOUNT(pBdb)));
-    if (pDfd[0] < samplesSize) {
+    uint32_t totalSize = sizeof(uint32_t) * (1 + KHR_DFDSIZEWORDS(KHR_DFDSAMPLECOUNT(pBdb)));
```

Comment:
This test is not doing anything hugely useful because `KHR_DFDSAMPLECOUNT` finds the number of samples by dividing the number of words in the descriptor block, `descriptorBlockSize`,  minus the `SAMPLESTART` offset by the number of words needed by a single sample. Therefore the reported number of samples is dependent on the data size. 

It is also not robust because it is possible for a descriptor to contain more than 1 descriptor block, though, we don't have any such today, each with its own `descriptorBlockSize`. `totalSize` is the total of all those blocks plus itself. The only way to know if you have more than one descriptor block is to compare `totalSize` with `descriptorBlockSize + 1`. If more there is another sample.

With the hardcoded +1 the test only ensures `totalSize` claims at least enough space to hold a Basic Descriptor Block with possibly 0 samples.

Given the data format specification I don't see any way to make a more robust test.

Please note the limitations of the test in a comment.

Reply:
More clearly and simply this test is exactly equivalent to

```
if (pDfd[0] < KHR_DFDVAL(pBdb, DESCRIPTORBLOCKSIZE) + 1)
```

Before using `KHR_DFDVAL` you should check that `pDfd[0]` is at least as large as `KHR_DF_WORD_SAMPLESTART` words so you know it won't crash.

This is a minimal sanity test given the possibility of multiple descriptor blocks. A full check requires looping over possible descriptor blocks (basically as long as pDfd[0] > 4) deducting the `descriptorBlockSize` of each descriptor found. As just noted care must be taken that there is enough data left when starting to process a descriptor such that `KHR_DFDVAL(pBdb, DESCRIPTORBLOCKSIZE)` can succeed without crashing. It should also verify there is at least 1 sample in each descriptor.

It is probably a good idea to add the full test.

Reply:
Thanks for the detailed explanation and pseudo-code, it makes sense to add more robust test here. I'm working on adding this test, however adding it causes few tests to faile
```
	354 - ktxToolsTest.info.invalid_file (Failed)
	409 - ktxToolsTest.validate.warning_6003_UnknownDFDBlock (Failed)
	411 - ktxToolsTest.validate.warning_6005_MultipleBDFD (Failed)
	412 - ktxToolsTest.validate.error_6006_DescriptorBlockSizeTooSmall (Failed)
	413 - ktxToolsTest.validate.error_6007_DescriptorBlockSizeTooBig (Failed)
	681 - ktxToolsTest.compare.ignore_dfd_all (Failed)
	684 - ktxToolsTest.compare.ignore_dfd_extended (Failed)
	685 - ktxToolsTest.compare.ignore_dfd_none (Failed)
	686 - ktxToolsTest.compare.ignore_bdfd_bytesplane (Failed)
```
Do you have some ideas how to tackle this?

Reply:
Thanks for the more robust test.

The CTS tests all compare the output from the program against golden files. I expect the more robust test is causing libktx to return a different error code and hence the tools to output different messages.

Use `ctest` to run the failing tests locally. Something like

```
ctest --test-dir build --output-on-failure -R ktxToolTests.validate
```

This will show you the output from the failed tests which will be the name of the file with the captured output and that of the golden file. You can then look at the files and see what is different.See [Conformance Test Suite](https://github.com/KhronosGroup/KTX-Software/blob/main/BUILDING.md#conformance-test-suite) for more information about setting up and running the tests from `ctest`.

If the captured output from the commands looks reasonable for the error condition you can regenerate the golden files by running the `clitest.py` command in `tests/cts/clitests`. You can also use this to run individual tests rather than using `ctest`. See [Regenerating Golden Files](https://github.com/KhronosGroup/KTX-Software-CTS/blob/main/README.md#regenerating-golden-files) for more information. Omitting `--regen-golden` from the example `clitest.py` command runs the test. You will see the same output as from `--output-on-failure`.

One you have regenerated the golden files you will need to make a PR in the [KTX-Software-CTS repo](https://github.com/KhronosGroup/KTX-Software-CTS) with the new tests. Make a new branch in `tests/cts` for the updated tests. I do not think you will be able to push this branch to KTX-Software-CTS so I think you will have to fork KTX-Software-CTS, push your new branch to your fork and make the PR from there.

Lastly you need to update the KTX-Software-CTS commit reference in KTX-Software and push it to this PR. You do this by committing `tests/cts` and pushing the result.

Reply:
Sorry for the delay, I regenerated test files and addressed the rest of the comments